image-source: Decrease slideshow source memory usage

cg2121 opened the pull request:
Now the slideshow source will only load 5 images at a time. Images are
loaded and deleted when necessary. Each image has a two image buffer.

kkartaltepe commented:
> I am mildly concerned that this will increase disk usage/wear 
Mostly a non-issue anyone who was able to open their entire slideshow will have enough unused ram to cache them all or at least a large portion typically.

I also agree there is no reason to introduce a change from random to shuffle in the same PR.

RytoEX commented:
> Mostly a non-issue anyone who was able to open their entire slideshow will have enough unused ram to cache them all or at least a large portion typically.

@kkartaltepe 
It looks like this behavior _only_ loads up to five images at any one time?  It doesn't try to load as many as it can.  Unless I've missed something?  From cg2121 in #development:
> Loads one image and a two image buffer. The other images are loaded when needed.

An explanation on the intended behavior might help.

To clarify my earlier opinion, I'd be more okay with the scope of these changes _if_ the PR title, PR description, and commit message indicated the scope.  I would still have concerns about changing behaviors for users and how to deal with those changes (user option, user knowledge, support knowledge, changelog, etc.).

kkartaltepe commented:
>only loads up to five images at any one time

Yes i mean in the OS disk cache is where they will be stored to prevent disk usage on the next load (in the case of random or getting through the end of the cycle). But if these people have enough free ram to load all their images into ram, then the savings after only loading a few should leave them with plenty of free ram for the OS to cache these which will hopefully survive until the next time they are used. (which is likely since the on disk forms are significantly smaller than the decoded images).

RytoEX commented:
@kkartaltepe 
Fair enough.

cg2121 commented:
Well I found the lines where the memory is leaking. Still don't how to fix it.

In the clear_buffer function:
`	if (next) {
		bfree(ss->files.array[0].path);
		da_erase(ss->files, 0);
	} else {
		bfree(ss->files.array[ss->files.num - 1].path);
		da_pop_back(ss->files);
	}`

cg2121 commented:
Nevermind, I have figured it out.

jp9000 commented on plugins/image-source/obs-slideshow.c:
```diff
@@ -861,15 +957,16 @@ static obs_properties_t *ss_properties(void *data)
 			const char *slash;
 
-			dstr_copy(&path, last->path);
+			dstr_copy(&path, path_);
```
The bstrdup/bfree feels unnecessary here.  Should just do:

```C
dstr_copy(&path, da_end(ss->paths));
```

cg2121 replied:
When I try your code I get a segmentation fault.

jp9000 commented:
I kind of agree with RytoEX that this shouldn't be changed from randomization to shuffle.  I feel like it should be able to randomize just as it did before.  It doesn't have to perform the randomization on the next upcoming item; but rather at least on the next item being loaded.

jp9000 commented:
I also feel like this should be an option.  I'm not sure though.  I think it would be fine for most use cases, but I'm not sure why people are using so many images in the slide show in the first place.  It actually boggles my mind that this is even a problem.  It was meant for streamers to show sponsors.  What the heck are people using this for that they have this problem?

RytoEX commented:
As far as I know, people are using it as a slideshow for hundreds of wallpaper images 1080p or higher.

jp9000 commented:
Hm, maybe making an option for it doesn't matter then.  Should probably just always be on.  The problem is I don't think image sources load asynchronously right now.  I'm a bit concerned about thread stalling, especially if people are using 1920x1080-sized images with it.

cg2121 commented:
I changed it back to random and fixed some bugs that I found.

SuslikV commented:
"Limit" instead of "Fix" in the naming. It was intended move to load all images into the RAM.
I think, that the amount of used RAM should be under the user control too (similar to Replay Buffer settings - limit the count of the images and the amount of used RAM).

A lot of things in obs has different styles and features (options and UI entries) - all that end user will use. Why not to reinvent your own properties window each time you need new features in the source? (last one is sarcasm, if not clear)

Techjar commented:
I think we should have both random and shuffle as an option. Personally I'd prefer shuffle, as random sometimes picks the same image very soon after it was last shown, which is a bit annoying.

Techjar commented on plugins/image-source/obs-slideshow.c:
```diff
@@ -208,6 +226,46 @@ static void add_file(struct slideshow *ss, struct darray *array,
 
+static void clear_buffer(struct slideshow *ss, bool next)
+{
+	if (ss->paths.num < 5 || !ss->paths.num || !ss->files.num)
```
I see this magic 5 used a lot. Perhaps there should be a constant for it instead?

jp9000 replied:
To clarify what Techjar is saying here, instead of using the number '5' throughout the code, you should create a macro at the top:
```C
#define MAX_LOADED 5
```
and then use that sort of macro throughout the codebase instead of using 5, both for the sake of code clarity and in case you want to change that number for whatever reason without having to change every instance where '5' is used (if it's used elsewhere).

jp9000 commented:
The random/shuffle options are not related to the PR -- for now its original random functionality is preferred for me.  I don't think there's a need to add shuffle right now, and I think it would be a waste of time/resources when there are more important things to be focusing on.

jp9000 commented on plugins/image-source/obs-slideshow.c:
```diff
@@ -208,6 +226,46 @@ static void add_file(struct slideshow *ss, struct darray *array,
+	int index;
+
+	if (ss->randomize)
+		index = random_file(ss);
```
Causes:
```
warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
```

jp9000 commented on plugins/image-source/obs-slideshow.c:
```diff
@@ -475,28 +565,30 @@ static void ss_next_slide(void *data)
 
-	do_transition(ss, false);
+	if (--ss->cur_item < 0 && !ss->randomize)
+		ss->cur_item = ss->paths.num - 1;
```
Also causes:
```
warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
```

cg2121 commented:
The issues are fixed now.

cg2121 commented:
I fixed the issue where it wouldn't work properly if there were less then MAX_LOADED images. The issue was also fixed where the images would resize when reloading when using random.

jp9000 commented:
One last change you should do before merging:

```diff
diff --git a/plugins/image-source/obs-slideshow.c b/plugins/image-source/obs-slideshow.c
index b29204572..f5b20ee29 100644
--- a/plugins/image-source/obs-slideshow.c
+++ b/plugins/image-source/obs-slideshow.c
@@ -958,15 +958,17 @@ static obs_properties_t *ss_properties(void *data)
 	if (ss) {
 		pthread_mutex_lock(&ss->mutex);
 		if (ss->paths.num) {
-			char *path_ = bstrdup(da_end(ss->paths));
+			char **p_last_path = da_end(ss->paths);
+			const char *last_path;
 			const char *slash;
 
-			dstr_copy(&path, path_);
+			last_path = p_last_path ? *p_last_path : "";
+
+			dstr_copy(&path, last_path);
 			dstr_replace(&path, "\\", "/");
 			slash = strrchr(path.array, '/');
 			if (slash)
 				dstr_resize(&path, slash - path.array + 1);
-			bfree(path_);
 		}
 		pthread_mutex_unlock(&ss->mutex);
 	}
```

da_end() in this case will return a pointer to a pointer (because the dynamic array is an array of character pointers, and da_end does not return the last item's value, but rather a pointer to the last item, do a `da_end` for a `DARRAY(char *)` will return a `char **`).  The bstrdup/bfree is unnecessary.  It was duplicating an invalid pointer.

jp9000 commented:
Great work.
